File: docs/JAMIA/comparisons/RPython/TM/7920688_join_key_type_mismatch.R
# ID: SO#7920688
# Language: R
# Bug class: Join
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: data.table join key type mismatch (int vs double). Join key types must match.
# Reproduction status: Live
# Type system catch: Join key types must match

library(data.table)

dt1 <- data.table(A1 = letters[1:10], B1 = sample(1:5, 10, replace = TRUE))
dt2 <- data.table(A2 = letters[c(1:5, 11:15)], B2 = sample(1:5, 10, replace = TRUE))

setkey(dt1, A1)
setkey(dt2, A2)

# which=TRUE returns positions with NA for non-matches
positions <- dt1[dt2, which = TRUE]
cat("Positions:", positions, "\n")

# BUG: negative indexing with NAs crashes
tryCatch(
  dt1[-positions],
  error = function(e) cat("Crash:", conditionMessage(e), "\n")
)

File: docs/JAMIA/comparisons/RPython/TM/7920688_join_key_type_mismatch.ts
/**
 * RPython SO#7920688 — non-joins with data.tables
 * Effect: Crash
 * Bug class: Nullable type
 *
 * In R data.table, which=TRUE returns NA for non-matching keys. Using these
 * NAs in negative indexing crashes. The user doesn't handle the NA positions.
 *
 * In tidy-ts, join operations produce typed DataFrames directly. There are
 * no implicit NA index positions to mishandle. Anti-join is done via filter.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

const dt1 = createDataFrame([
  { a: "a", b: 1 },
  { a: "b", b: 4 },
  { a: "c", b: 2 },
  { a: "f", b: 2 },
  { a: "g", b: 3 },
]);

const dt2Keys = new Set(["a", "b", "k"]);

// Anti-join via filter — explicit, no NA confusion
const antiJoin = dt1.filter((r) => !dt2Keys.has(r.a));
antiJoin.print();

// Type system catches passing string key column to numeric ops
// @ts-expect-error — string[] is not assignable to number[]
const wrong = s.sum(dt1.extract("a"));
